[JIT] Separate torchbind from Python

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29501 [JIT] Separate torchbind from Python**

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

Commit:
[WIP][JIT] Separate torchbind from Python

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"




[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue


[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue


[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue


[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Separate torchbind from Python"


Currently blocked on schema serialization issue

Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/python_custom_class.cpp:
```diff
@@ -0,0 +1,71 @@
+      ctor_wrapper = &classCU()->get_function(wrapper_fn_name);
+    } else {
+      auto graph = std::make_shared<Graph>();
+      ctor_wrapper = classCU()->create_function(wrapper_fn_name, graph);
```
FunctionSchema is not registered so keyword argument handling for init will fail.

zdevito commented on torch/csrc/jit/python_custom_class.cpp:
```diff
@@ -0,0 +1,71 @@
+void initPythonCustomClassBindings(PyObject* module) {
+  auto m = py::handle(module).cast<py::module>();
+
+  m.def("_get_custom_class_python_wrapper", [](const std::string& str) {
```
this really needs a comment, I spent 5 minutes trying to figure out what it does.

zdevito replied:
Also probably misnamed. It retries the python wrapper for the _constructor_ of the class, not the class itself.

jamesr66a replied:
The analogue to Python is the code object, so that's why I called it class wrapper and not constructor wrapper. I can still change it if you want

zdevito commented on torch/csrc/jit/script/schema_type_parser.cpp:
```diff
@@ -214,6 +215,29 @@ std::pair<TypePtr, c10::optional<AliasInfo>> SchemaTypeParser::parseType() {
+    }
+    L.expect('.');
+    auto class_tok = L.expect(TK_IDENT);
+    value = getCustomClass(
```
We should write up a design for how namespaces work with custom classes. I think we want a unified namespace where user-defined classes live in the same namespace as everything else in the schema. Demoting them to `__torch__.torch.classes.` is weird because there is nothing valid in `__torch__` or `__torch__.torch`, and because it is going to be confusing to have a user defined class called Dict that is not the same class as Dict.

zdevito commented on torch/custom_class_detail.h:
```diff
@@ -0,0 +1,21 @@
+namespace detail {
+
+template <class RetType, class...>
+struct types {
```
I looked into this, and I believe its use is not necessary, along with a bunch of heavily templated functions in custom_class.h (addInputs, in particular). We want to minimize template meta-programming locations. Once `    registeredOps.push_back(
        torch::RegisterOperators().op(qualFuncName, std::move(func)));` has been called, the program can simply look up all the information it needs from that operators schema.

jamesr66a replied:
I've gone through and deleted what I can, but I believe these are still necessary to recover the argument types to instantiate the wrapper lambdas

jamesr66a replied:
I've just deleted a bunch of template stuff just now, but I believe this in particular is still needed to separate out argument types from the Func type to create the wrapper lambdas in `def_()`

zdevito commented on test/test_jit.py:
```diff
@@ -4329,6 +4329,34 @@ def forward(self, x):
+
+        with self.assertRaisesRegex(RuntimeError, "Expected a value of type 'int'"):
+            val = torch.classes._TorchScriptTesting_Foo(5, 3)
+            val.increment('foo')
```
I don't see the bindings for Object, I am assuming they are already a part of a previous PR.

zdevito reviewed (commented):
Looks pretty good as a way to get python out. Comments inline.

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 2cc8edd:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2cc8edde412ababd252f797f8cc04c401fe6f461).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3936758) pytorch_linux_backward_compatibility_check_test (1/1)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=73682255))
```
Dec 13 01:00:00 +++ eval 'extract_trap_cmd ' 
Dec 13 01:00:00 ++++ extract_trap_cmd 
Dec 13 01:00:00 ++++ printf '%s\n' '' 
Dec 13 01:00:00 +++ printf '%s\n' cleanup 
Dec 13 01:00:00 ++ trap -- ' 
Dec 13 01:00:00 cleanup' EXIT 
Dec 13 01:00:00 ++ which sccache 
Dec 13 01:00:00 ++ sccache --stop-server 
Dec 13 01:00:00 Stopping sccache server... 
Dec 13 01:00:00 error: couldn't connect to server 
Dec 13 01:00:00 caused by: Connection refused (os error 111) 
Dec 13 01:00:00 ++ true 
Dec 13 01:00:00 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 13 01:00:00 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 13 01:00:00 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 13 01:00:00 ++ RUST_LOG=sccache::server=error 
Dec 13 01:00:00 ++ sccache --start-server 
Dec 13 01:00:00 Starting sccache server... 
Dec 13 01:00:00 ++ sccache --zero-stats 
Dec 13 01:00:00 Compile requests                 0 
Dec 13 01:00:00 Compile requests executed        0 
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

houseroad reviewed (commented):
seems the schema parser doesn't support parsing this type (__torch__.torch.classes._TorchScriptTesting_StackString) well. Could you please enable the support for this case?

Dec 13 02:01:24 Traceback (most recent call last):
Dec 13 02:01:24   File "check_backward_compatibility.py", line 86, in <module>
Dec 13 02:01:24     s = parse_schema(line.strip())
Dec 13 02:01:24 RuntimeError: 
Dec 13 02:01:24 unknown type specifier:
Dec 13 02:01:24 _TorchScriptTesting_StackString::pop(__torch__.torch.classes._TorchScriptTesting_StackString _0) -> (str _0)
Dec 13 02:01:24                                      ~~~~~~~~~ <--- HERE

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

zdevito commented on torch/csrc/jit/pybind_utils.h:
```diff
@@ -696,9 +686,8 @@ inline py::object toPyObject(IValue ivalue) {
-    auto res = tryToConvertToCustomClass(obj);
-    if (res.has_value()) {
-      return res.value();
+    if (obj->name().find("__torch__.torch.classes") == 0) {
```
Now that we can bind directly to py::Object should we just universally return one rather than doing the copy below?

zdevito reviewed (approved):
It'd be nice to come up with a plan for name mangling of custom classes, but the code looks good and this will unblock a lot of things.

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

houseroad commented:
If not possible to reproduce the problem on local build, we may want to try docker then :-)

suo commented:
@pytorchbot rebase this please

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

jamesr66a commented:
@pytorchbot rebase this please

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

Commit:
Update on "[JIT] Separate torchbind from Python"


Differential Revision: [D18463063](https://our.internmc.facebook.com/intern/diff/D18463063)

[ghstack-poisoned]

houseroad reviewed (commented):
After some thoughts, I think the culprit is that when doing the backward compatibility check we use the nightly build, which doesn't contain the support for the new type introduced in this PR. 
To mitigate the problem, we should skip the new op here: https://github.com/pytorch/pytorch/blob/master/test/backward_compatibility/check_backward_compatibility.py#L85
